lib/docker-file.ts

```diff
@@ -26,10 +26,45 @@ async function readDockerfileAndAnalyse(
   return analyseDockerfile(contents);
 }
 
+function resolveVariables(
+  text: string,
+  instruction: Instruction,
+  dockerfile: Dockerfile,
+): string {
+  const variables = instruction.getVariables();
+  if (variables) {
+    const resolvedVariables = variables.reduce((resolvedVars, variable) => {
+      const line = variable.getRange().start.line;
+      const name = variable.getName();
+      resolvedVars[name] = dockerfile.resolveVariable(name, line);
+      return resolvedVars;
+    }, {});
+    Object.keys(resolvedVariables).forEach((variable) => {
+      text = text.replace(`\$\{${variable}\}`, resolvedVariables[variable]);
+    });
+  }
+  return text;
+}
+
 async function analyseDockerfile(
   contents: string,
 ): Promise<DockerFileAnalysis | undefined> {
   const dockerfile = DockerfileParser.parse(contents);
+  // collect stages names
+  const stages = dockerfile.getFROMs().reduce((names, from) => {
```

Comment:
This reduce is getting rather long - you might want to consider moving this into a separate function which returns the stages.

Reply:
Done, created a new function for base image retrieval https://github.com/snyk/snyk-docker-plugin/pull/98/files#diff-138b9629a01f1781af8ad6dff1ea75acR109